File: ALE.cpp
#include<iostream>
#include<stdio.h>
#include<string.h>

using namespace std;

struct encabezado
{
    int NRS, PR, URE;
}e;

struct registro
{
    int NR, SR, ARE;
    char nom[20], estado[20];
}n,a,s,r;

FILE *fd;
int le = sizeof(struct encabezado);
int lr = sizeof(struct registro);

void escribir()
{
    char rpta; int sgte,pos; bool band, nomRepetido;
    e.NRS = 0; e.PR = -1; e.URE = -1, n.NR = 0;

    if ((fd = fopen("personasALE.txt","w+t")) == NULL)
    {
        cout << "No se pudo crear el archivo." << endl;
        return;
    }
    
    fwrite(&e, le, 1, fd);

    do
    {
        fflush(stdin);

        cout << "\nTarea "<< n.NR + 1 << endl;
        cout << "Nombre: "; cin >> n.nom;
        cout << "Estado: "; cin >> n.estado;   
        n.ARE = 0; 
        sgte = e.PR; nomRepetido = false;

        while (sgte != -1)
        {
            pos = (sgte - 1) * lr + le;
            fseek(fd, pos, 0);
            fread(&s, lr, 1, fd);

            if (strcmp(n.nom, s.nom) == 0)
            {
                nomRepetido = true;
                cout << "Ya existe un registro con el mismo nombre. Ingrese otro nombre." << endl;
                break;
            }

            sgte = s.SR;
        }

        if (nomRepetido == false)
        {
            n.NR = ++e.NRS;
            if (e.PR == -1)
            {
               n.SR = e.PR;
                e.PR = n.NR;
                fwrite(&n, lr, 1, fd);
            }
            else
            {
                sgte = e.PR; band = false;

                while (sgte != -1)
                {
                    pos = (sgte-1) * lr + le;
                    fseek(fd, pos, 0);
                    fread(&s, lr, 1, fd);

                    if (strcmp(n.nom, s.nom) > 0) 
                    {
                        band = true;
                        a = s;
                        sgte = s.SR;
                        continue;
                    }
                    break;
                }
            
                if (band == false)
                {
                    n.SR = e.PR;
                    e.PR = n.NR;
                }
                if (band == true)
                {
                    n.SR = a.SR;
                    a.SR = n.NR;

                    pos = (a.NR-1) * lr + le;
                    fseek(fd, pos, 0);
                    fwrite(&a, lr, 1, fd);
                }
            
                pos = (n.NR-1) * lr + le;
                fseek(fd, pos, 0);
                fwrite(&n, lr, 1, fd);
            }
        }
        
        cout << "Desea mas registros? (s/n)" <<endl; cin >> rpta;       
    } while (rpta == 's' || rpta == 'S');

    fseek(fd, 0, 0);
    fwrite(&e, le, 1, fd);
    fclose(fd);
}

void mostrar()
{
    int sgte,pos;

    if ((fd=fopen("personasALE.txt","rt")) == NULL)
    {
        cout << "No se puedo abrir el archivo" << endl;
        return;
    }

    fread(&e, le, 1, fd);
    sgte = e.PR;

    cout << "\nTAREAS:" <<endl;
    while (sgte != -1)
    {
        pos = (sgte - 1) * lr + le;
        fseek(fd, pos, 0);
        fread(&s, lr, 1, fd);
        cout << "Nombre: " << s.nom << ", Estado: " << s.estado << endl;
        sgte = s.SR;
    }
    
    fclose(fd);
}

void agregar()
{
    char rpta; int sgte,pos; bool band, nomRepetido;

    if ((fd=fopen("personasALE.txt","r+t")) == NULL)
    {
        cout << "No se pudo crear el archivo" <<endl;
        return;
    }

    fread(&e, le, 1, fd);

    do
    {
        fflush(stdin);
        cout << "\nNombre: "; cin >> n.nom;
        cout << "Estado: "; cin >> n.estado;
        n.ARE = 0;
        sgte = e.PR; nomRepetido = false;

        while (sgte != -1)
        {
            pos = (sgte - 1) * lr + le;
            fseek(fd, pos, 0);
            fread(&s, lr, 1, fd);

            if (strcmp(n.nom, s.nom) == 0)
            {
                nomRepetido = true;
                cout << "Ya existe un registro con el mismo nombre. Ingrese otro nombre." << endl;
                break;
            }

            sgte = s.SR;
        }

        if (nomRepetido == false)
        {
            n.NR = ++e.NRS;

            if (e.PR == -1)
            {
                n.SR = e.PR;
                e.PR = n.NR;
                pos = (n.NR - 1) * lr + le;
                fseek(fd, pos, 0);
                fwrite(&n, lr, 1, fd);
            }
            else
            {
                sgte = e.PR;

                while (sgte != -1)
                {
                    pos = (sgte - 1) * lr + le;
                    fseek(fd, pos, 0);
                    fread(&s, lr, 1, fd);
                    
                    if (strcmp(n.nom,s.nom) > 0)
                    {
                        band = true;
                        a = s;
                        sgte = s.SR;
                        continue;
                    }
                    break;
                }
            
                if (band == false)
                {
                    n.SR = e.PR;
                    e.PR = n.NR;
                }
            
                if (band == true)
                {
                    n.SR = a.SR;
                    a.SR = n.NR;
                    pos = (a.NR - 1) * lr + le;
                    fseek(fd, pos, 0);
                    fwrite(&a, lr, 1, fd);
                }
            
                pos = (n.NR - 1) * lr + le;
                fseek(fd, pos, 0);
                fwrite(&n, lr, 1, fd);
            }
        }
        
        cout << "Desea mas registros? (s/n)" <<endl; cin >> rpta; 
    } while (rpta == 's' || rpta == 'S');
    
    fseek(fd, 0, 0);
    fwrite(&e, le, 1, fd);
    fclose(fd);
}

void eliminar()
{
    char rpta; int sgte,pos; bool band,flag=false;
    char nom_eliminado[20];

    if ((fd=fopen("personasALE.txt","r+t")) == NULL)
    {
        cout << "No se pudo abrir el archivo" << endl;
        return;
    }
    
    fread(&e, le, 1, fd);
    fflush(stdin);
    cout << "\nNombre: "; cin >> nom_eliminado;

    while (fread(&r, lr, 1, fd) == 1)
    {
        if(strcmp(nom_eliminado,r.nom) == 0){
            flag = true;
            r.ARE = e.URE;
            e.URE = r.NR;
            break;
        }
    }
    
    if (flag == false)
    {
        cout << "El nombre no existe." << endl;
        return;
    }

    sgte = e.PR; band = false;

    while (sgte != -1)
    {
        pos = (sgte - 1) * lr + le;
        fseek(fd, pos, 0);
        fread(&s, lr, 1, fd);

        if (strcmp(r.nom,s.nom) > 0)
        {
            band = true;
            a = s;
            sgte = s.SR;
            continue;
        }
        break;
    }
    
    if (band == false)
    {
        e.PR = r.SR;
    }

    if (band == true)
    {
        a.SR = r.SR;
        pos = (a.NR - 1) * lr + le;
        fseek(fd, pos, 0);
        fwrite(&a, lr, 1, fd);

        cout << "Registro Eliminado" << endl;
    }
    
    pos = (r.NR - 1) * lr + le;
    fseek(fd, pos, 0);
    fwrite(&r, lr, 1, fd);
    fseek(fd, 0, 0);
    fwrite(&e, le, 1, fd);
    fclose(fd);
}

void buscar()
{
    char nombre[20]; int sgte, pos; bool encontrado = false;

    fflush(stdin);
    cout << "\nIngrese el nombre a buscar: "; cin >> nombre;

    if ((fd=fopen("personasALE.txt","rt")) == NULL)
    {
        cout << "No se pudo abrir el archivo" << endl;
        return;
    }

    fread(&e, le, 1, fd);
    sgte = e.PR;

    while (sgte != -1)
    {
        pos = (sgte -1) * lr + le;
        fseek(fd, pos, 0);
        fread(&s, lr, 1, fd);

        if (strcmp(s.nom,nombre) == 0)
        {
            encontrado = true;
            break;
        }
        
        a = s;
        sgte = s.SR;
    }
    
    if (encontrado == true)
    {
        cout << nombre << " ha sido encontrado" << endl;
        cout << "Estado: " << s.estado << endl;
    }
    else
    {
        cout << nombre << " no ha sido encontrado" << endl;
    }
    
    fclose(fd);
}

void modificar()
{
    char nombre[20]; int sgte, pos; bool encontrado = false;

    cout << "\nIngrese el nombre de la tarea: "; cin >> nombre;

    if ((fd = fopen("personasALE.txt", "r+t")) == NULL)
    {
        cout << "No se pudo abrir el archivo" << endl;
        return;
    }

    fread(&e, le, 1, fd);
    sgte = e.PR;

    while (sgte != -1)
    {
        pos = (sgte - 1) * lr + le;
        fseek(fd, pos, 0);
        fread(&s, lr, 1, fd);

        if (strcmp(s.nom, nombre) == 0)
        {
            encontrado = true;
            break;
        }

        sgte = s.SR;
    }

    if (encontrado)
    {
        cout << "Nombre: " << s.nom << ", Estado: " << s.estado << endl;
        cout << "Ingrese el nuevo estado: ";
        cin >> s.estado;

        pos = (s.NR - 1) * lr + le;
        fseek(fd, pos, 0);
        fwrite(&s, lr, 1, fd);

        cout << "Registro modificado exitosamente." << endl;
    }
    else
    {
        cout << nombre << " no ha sido encontrado" << endl;
    }

    fclose(fd);
}

void ver_archivo()
{
    if ((fd = fopen("personasALE.txt","rt")) == NULL)
    {
        cout << "No se puedo abrir el archivo." << endl;
        return;
    }

    fread(&e, le, 1, fd);
    cout << "\nNRS: " << e.NRS << "\tPR: " << e.PR << "\tURE: " << e.URE << endl; 
    cout << "NR\tNombre\tEstado\tSR\tARE" << endl;

    while (fread(&s, lr, 1, fd) == 1)
    {
        cout << s.NR << "\t" << s.nom << "\t" << s.estado << "\t" << s.SR << "\t" << s.ARE << endl;
    }
}

int main(int argc, char *argv[])
{
    int opc = 0;

    do
    {
        cout << "\nMENU"; 
        cout << "\n1. Escribir Tareas";
        cout << "\n2. Mostrar Tareas";
        cout << "\n3. Agregar Tarea";
        cout << "\n4. Eliminar Tarea";
        cout << "\n5. Buscar Tarea";
        cout << "\n6. Modificar Tarea";
        cout << "\n7. Ver Archivo";
        cout << "\n8. Salir";
        cout << "\nSeleecione una opcion: "; cin >> opc;

        switch (opc)
        {
        case 1:
            escribir();
            break;

        case 2:
            mostrar();
            break;

        case 3:
            agregar();
            break;

        case 4:
            eliminar();
            break;

        case 5:
            buscar();
            break;

        case 6:
            modificar();
            break;

        case 7:
            ver_archivo();
            break;
    
        default:
            break;
        }
    } while (opc != 8);

    return 0;
}

File: ALE.py
class Encabezado:
    def __init__(self):
        self.NRS = 0
        self.PR = -1
        self.URE = -1

class Registro:
    def __init__(self, NR, nom, estado):
        self.NR = NR
        self.nom = nom
        self.estado = estado
        self.SR = -1
        self.ARE = 0

def escribir():
    registros = []
    encabezado = Encabezado()

    try:
        with open("personasALEP.txt", "w") as file:
            while True:
                NR = encabezado.NRS + 1
                nom = input("Nombre: ")
                estado = input("Estado: ")

                nuevo_registro = Registro(NR, nom, estado)

                if encabezado.PR == -1:
                    nuevo_registro.SR = encabezado.PR
                    encabezado.PR = NR
                else:
                    sgte = encabezado.PR
                    anterior = None

                    while sgte != -1 and nuevo_registro.nom > registros[sgte - 1].nom:
                        anterior = sgte
                        sgte = registros[sgte - 1].SR

                    nuevo_registro.SR = sgte

                    if anterior is not None:
                        registros[anterior - 1].SR = NR
                    else:
                        encabezado.PR = NR

                registros.append(nuevo_registro)
                encabezado.NRS += 1

                respuesta = input("Desea más registros? (s/n): ")
                if respuesta.lower() != 's':
                    break

            # Escribir el encabezado actualizado al archivo
            file.write(f"{encabezado.NRS} {encabezado.PR} {encabezado.URE}\n")

            # Escribir registros al archivo
            for registro in registros:
                file.write(f"{registro.NR} {registro.nom} {registro.estado} {registro.SR} {registro.ARE}\n")

    except IOError:
        print("No se pudo crear el archivo.")

def eliminar():
    try:
        with open("personasALEP.txt", "r") as file:
            lines = file.readlines()

        # Leer el encabezado
        encabezado_data = lines[0].split()
        encabezado = Encabezado()
        encabezado.NRS, encabezado.PR, encabezado.URE = map(int, encabezado_data)

        # Obtener el nombre de la tarea a eliminar
        nom_eliminado = input("Nombre de la tarea a eliminar: ")

        # Buscar el registro a eliminar y actualizar la lista de registros
        registros = []
        flag = False
        for line in lines[1:]:
            data = line.split()
            registro = Registro(int(data[0]), data[1], data[2])
            registro.SR, registro.ARE = map(int, data[3:5])

            if registro.nom == nom_eliminado:
                flag = True
                registro.ARE = encabezado.URE
                encabezado.URE = registro.NR
            registros.append(registro)

        if not flag:
            print("La tarea no existe.")
            return

        # Verificar si la tarea eliminada es la primera y actualizar PR
        if registros and registros[0].nom == nom_eliminado:
            encabezado.PR = registros[0].SR

        # Escribir la información actualizada al archivo
        with open("personasALEP.txt", "w") as file:
            file.write(f"{encabezado.NRS} {encabezado.PR} {encabezado.URE}\n")
            for registro in registros:
                file.write(f"{registro.NR} {registro.nom} {registro.estado} {registro.SR} {registro.ARE}\n")

        print("Tarea eliminada con éxito.")

    except IOError:
        print("No se pudo abrir o modificar el archivo.")


def ver_archivo():
    try:
        with open("personasALEP.txt", "r") as file:
            line = file.readline().split()
            encabezado = Encabezado()
            encabezado.NRS, encabezado.PR, encabezado.URE = map(int, line)
            
            print(f"NRS: {encabezado.NRS}\tPR: {encabezado.PR}\tURE: {encabezado.URE}")
            print("NR\tNombre\tEstado\tSR\tARE")

            for line in file:
                data = line.split()
                registro = Registro(int(data[0]), data[1], data[2])
                registro.SR, registro.ARE = map(int, data[3:5])
                print(f"{registro.NR}\t{registro.nom}\t{registro.estado}\t{registro.SR}\t{registro.ARE}")

    except IOError:
        print("No se pudo abrir el archivo.")

def main():
    escribir()
    ver_archivo()
    eliminar()
    ver_archivo()

if __name__ == "__main__":
    main()